modules/aui2s/aui2s_src.c

```diff
@@ -0,0 +1,170 @@
+/**
+ * @file aui2s_src.c freeRTOS I2S audio driver module - recorder
+ *
+ * Copyright (C) 2019 cspiel.at
+ */
+#include <sys/types.h>
+#include <sys/time.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <pthread.h>
+#include "freertos/FreeRTOS.h"
+#include "driver/i2s.h"
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+#include "aui2s.h"
+
+
+struct ausrc_st {
+	const struct ausrc *as;  /* pointer to base-class (inheritance) */
+	pthread_t thread;
+	bool run;
+	void *sampv;
+	size_t sampc;
+	size_t bytes;
+	ausrc_read_h *rh;
+	void *arg;
+	struct ausrc_prm prm;
+
+	uint32_t *pcm;
+};
+
+
+static void ausrc_destructor(void *arg)
+{
+	struct ausrc_st *st = arg;
+
+	/* Wait for termination of other thread */
+	if (st->run) {
+		info("aui2s: stopping recording thread\n");
+		st->run = false;
+		(void)pthread_join(st->thread, NULL);
+	}
+
+	mem_deref(st->sampv);
+	mem_deref(st->pcm);
+}
+
+
+static void convert_pcm(struct ausrc_st *st, size_t i, size_t n)
+{
+	uint32_t j;
+	uint16_t *sampv = st->sampv;
+	for (j = 0; j < n; j++) {
+		uint32_t v = st->pcm[j];
+		uint16_t *o = sampv + i + j;
+		*o = v >> 15;
+		/* if negative fill with ff */
+		if (v & 0x80000000)
+			*o |= 0xfffe0000;
+	}
+}
+
+
+static void *read_thread(void *arg)
+{
+	struct ausrc_st *st = arg;
+	int err;
+
+	err = aui2s_start(st->prm.srate, I2O_RECO);
```

Comment:
is the I2S started from the thread on purpose ?

or can it be started from the alloc function?

(if aui2s_start fails, the thread will just exit but the program will continue
as if everything is working).